src/Conversion/ONNXToKrnl/NN/Normalization.cpp

```diff
@@ -383,10 +495,356 @@ struct ONNXLayerNormalizationOpLowering
+                    [&](SCFBuilder &scf, Value blockLocalInd) {
+                      MDBuilder create(scf);
+                      generateIterWithSIMD(rewriter, create, lnOp, XFlatMemRef,
+                          scaleFlatMemRef, BFlatMemRef, YFlatMemRef,
+                          meanFlatMemRef, invStdDevFlatMemRef, tmpRedMemRef,
+                          tmpRedMemRef2, XFlatDims[1], blockLocalInd, epsilon,
+                          1, VL, hasScalarScale, hasScalarB);
+                    }); /* for inside blocked loop */
+              },
+              [&](SCFBuilder &scf) {
+                MDBuilder create(scf);
+                // create.krnl.printf("full tile\n");
+                generateIterWithSIMD(rewriter, create, lnOp, XFlatMemRef,
+                    scaleFlatMemRef, BFlatMemRef, YFlatMemRef, meanFlatMemRef,
+                    invStdDevFlatMemRef, tmpRedMemRef, tmpRedMemRef2,
+                    XFlatDims[1], blockedLoopIndices[0], epsilon, B, VL,
+                    hasScalarScale, hasScalarB);
+              });
+        }); /* blocked loop */
+    // We don't seem to need to reshape fhe flattened memref, just pass the
+    // original ones.
+    replaceLayerNormalizationOp(
+        rewriter, lnOp, YMemRef, meanMemRef, invStdDevMemRef);
+    return success();
+  }
 };
 
+/*
+ *  Explanation on the SIMD code.
+
+ * Algorithm.
+
+  There are 2 ways to compute var[x]. The first is the method listed in the op
+  definition:
+
+  E[ (x -E[x]) ^2]  where E[x] corresponds to the mean of x.
+
+  The second method is:
+
+  E[x]^2 - E[x^2].
+
+  As shown in wikipedia (https://en.wikipedia.org/wiki/Variance)
+
+   E[ (x -E[x]) ^2]) = E[ x^2 -2xE[x] + E[x]^2 ] =
+   E[x^2] -2E[x]E[x] + E[x]^2 = E[x^2] - E[x]^2
+
+  This second method is faster as we can compute both E[x] and E[x^2] at the
+  same time.
+
+
+ * Python code that explains how the code was derived.
+
+  def layer_norm_simd2_v3(x, a, scale, b):
+    (b1, b2) = (2, 4)
+    t_rank = len(x.shape)
+    assert t_rank==2, "work for rank 2 only"
+    (a1, s2)  = x.shape
+    mean = np.zeros((a1, 1))
+    inv_std_dev = np.zeros((a1, 1))
+    y = np.zeros((a1, s2))
+    for i1 in range(0, a1, b1):
+        # iterate over blocks of b1 values
+
+        # MEAN(x), MEAN(x2)
+        # iterate over a_block, s_block: parallel add
+        r = np.zeros((b1, b2))
+        r_square = np.zeros((b1, b2))
+        for i2 in range(0, s2, b2): # Unroll B1, SIMD by B2,
+            xx = x[i1:i1+b1, i2:i2+b2]
+            xxx = np.multiply(xx, xx)
+            r = np.add(r, xx)
+            r_square = np.add(r_square, xxx)
+
+        # simd reduction; res B1 x 1
+        # 2 B1 div
+        mean_b = np.sum(r, axis=1, keepdims=True) # SIMD reduction to (B1x1)
+          # values. mean_b = np.divide(mean_b, s2) # (B2x1) values... so scalar
+          # is ok. mean_square_b = np.sum(r_square, axis=1, keepdims=True)
+          # Same.
```

Comment:
This part seems to be broken due to clang-format. The same for the below code for `normalized_scaled_b`. You can use `clang-format on/off` to keep the python code unmodified.

Reply:
Will do